### PR TITLE
Fixing MyPy issues inside providers/microsoft

### PR DESCRIPTION
**Pull request by khalidmammadov**
Part of https://github.com/apache/airflow/issues/19891

Issue fixed:
- Ignoring issue when mandatory arguments are not provided but provided from default_args instead
- Fixing type signature for `_read` method of `WasbTaskHandler`
- Fixing type of execute method's context parameter as it currently violates (type wise)  Liskov principle from SOLID, due to different signatures from base class

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
@khalidmamadov :  There is one problem. Context does not exist in Airlfow 2.1 so you will have to change it to Any - until we decide to make providers  depends on 2.3+

**Comment by khalidmammadov** [comment]
> @khalidmamadov : There is one problem. Context does not exist in Airlfow 2.1 so you will have to change it to Any - until we decide to make providers depends on 2.3+

I see... let me fix

**Comment by josh-fell** [comment]
@potiuk Since some of the example DAGs (in this provider and several others) that have Mypy complaining about `default_args`, WDYT about adding to the logic when we build docs that ignores the `# type: ignore...` comments similar to what we do with the `# START/END [...]` markers so they don't show up in documentation? In this provider for example, we hide the Mypy ignore comment in "example_azure_blob_to_gcs.py".

**Comment by uranusjr** [comment]
Context is a TypedDict, which naturally downcasts to `Dict[str, Any]`.

**Comment by potiuk** [comment]
> Context is a TypedDict, which naturally downcasts to `Dict[str, Any]`.

Oh yeah. that would be the best soluition - using `Dict[str,Any]` instead of Context should work also for 2.1 compatibility!

**Comment by potiuk** [comment]
> @potiuk Since some of the example DAGs (in this provider and several others) that have Mypy complaining about `default_args`, WDYT about adding to the logic when we render docs that ignores the `# type: ignore...` comments similar to what we do with the `# START/END [...]` markers so they don't show up in documentation? In this provider for example, we hide the Mypy ignore comment in "example_azure_blob_to_gcs.py".

Should be possible. I think for that we would have to write a little extra logic here: https://github.com/apache/airflow/blob/f9591042285e8acefebdab0f789357049e136f64/docs/exts/exampleinclude.py#L52

**Comment by khalidmammadov** [comment]
> Context is a TypedDict, which naturally downcasts to `Dict[str, Any]`.

I will try as you suggest

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
> > Context is a TypedDict, which naturally downcasts to `Dict[str, Any]`.
> 
> I will try as you suggest

It will not work as exepected because of Liskov substitution principle. 

```
airflow/providers/apache/hive/sensors/metastore_partition.py:70: error: Argument 1 of "poke" is incompatible with supertype "BaseSensorOperator"; supertype defines the argument type as "Context"
        def poke(self, context: MutableMapping[str, Any]) -> Any:
        ^
airflow/providers/apache/hive/sensors/metastore_partition.py:70: note: This violates the Liskov substitution principle
airflow/providers/apache/hive/sensors/metastore_partition.py:70: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

But I think I found a nice "workaround":

In imports section of all operators/sensors:

```
if TYPE_CHECKING:
    from airflow.utils.context import Context
else:
    Context = MutableMapping[str, Any]
```

**Comment by potiuk** [comment]
Example on how we can treat Context here: #20422

**Comment by khalidmammadov** [comment]
> > > Context is a TypedDict, which naturally downcasts to `Dict[str, Any]`.
> > 
> > 
> > I will try as you suggest
> 
> It will not work as exepected because of Liskov substitution principle.
> 
> ```
> airflow/providers/apache/hive/sensors/metastore_partition.py:70: error: Argument 1 of "poke" is incompatible with supertype "BaseSensorOperator"; supertype defines the argument type as "Context"
>         def poke(self, context: MutableMapping[str, Any]) -> Any:
>         ^
> airflow/providers/apache/hive/sensors/metastore_partition.py:70: note: This violates the Liskov substitution principle
> airflow/providers/apache/hive/sensors/metastore_partition.py:70: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
> ```
> 
> But I think I found a nice "workaround":
> 
> In imports section of all operators/sensors:
> 
> ```
> if TYPE_CHECKING:
>     from airflow.utils.context import Context
> else:
>     Context = MutableMapping[str, Any]
> ```

This look good, will try this as well

File: airflow/providers/microsoft/azure/log/wasb_task_handler.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def close(self) -> None:
         # Mark closed so we don't double write if close is called twice
         self.closed = True
 
-    def _read(self, ti, try_number: str, metadata: Optional[str] = None) -> Tuple[str, Dict[str, bool]]:
```
Comment (potiuk) [inline airflow/providers/microsoft/azure/log/wasb_task_handler.py:110]: hmmmmmm :)

**Review by potiuk — APPROVED** [review]
I wish we had better solution for the "default_args" but it must do for now :(

**Comment by potiuk** [comment]
Some static checks :(

File: airflow/providers/microsoft/azure/log/wasb_task_handler.py (hunk 1 above)
Comment (khalidmammadov) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py:110]: Is there anything wrong here? didn't get :)

**Comment by khalidmammadov** [comment]
> Some static checks :(

Yes, pre-commit was disabled... re-enabled now and fixed

**Comment by potiuk** [comment]
> > Some static checks :(
> 
> Yes, pre-commit was disabled... re-enabled now and fixed

BTW. You now you can disable just selected pre-commits @khalidmammadov :

```
export SKIP="mypy"
```

Also `git commit --no-verify` will skip pre-commit just for this one commut without disabling it for the repo.

:D

**Comment by khalidmammadov** [comment]
> > > Some static checks :(
> > 
> > 
> > Yes, pre-commit was disabled... re-enabled now and fixed
> 
> BTW. You know you can disable just selected pre-commits @khalidmammadov ?
> 
> ```
> export SKIP="mypy"
> ```
> 
> Also `git commit --no-verify` will skip pre-commit just for this one commit without disabling it for the repo.
> 
> :D

Ah, didnt know that (looked up for something like that but didn't find..) thank for that!

**Comment by potiuk** [comment]
> Ah, didnt know that (looked up for something like that but didn't find..) thank for that!

You are most welcome. Pre-commit has a number of useful things built in - for example you can easily run pre-commit on the last commit only:

```
pre-commit run --from-ref=HEAD^ --to-ref=HEAD
```

Also see here: https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst#id1 "using pre-commit" for some other stuff.

**Review by potiuk — APPROVED** [review]

**Comment by khalidmammadov** [comment]
> > Ah, didnt know that (looked up for something like that but didn't find..) thank for that!
> 
> You are most welcome. Pre-commit has a number of useful things built in - for example you can easily run pre-commit on the last commit only:
> 
> ```
> pre-commit run --from-ref=HEAD^ --to-ref=HEAD
> ```
> 
> Also see here: https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst#id1 "using pre-commit" for some other stuff.

Thanks for these, I will definitely will make use of these!

**Pull request merged**